tests/src/DeprecationRulesTest.php

```diff
@@ -47,5 +47,17 @@ public function dataDeprecatedSamples(): \Generator
                 'Call to deprecated constant DATETIME_DATE_STORAGE_FORMAT: Deprecated in drupal:8.5.0 and is removed from drupal:9.0.0. Use \Drupal\datetime\Plugin\Field\FieldType\DateTimeItemInterface::DATE_STORAGE_FORMAT instead.',
             ]
         ];
+        yield [
+            __DIR__ . '/../fixtures/drupal/core/modules/simpletest/src/TestDiscovery.php',
```

Comment:
This highlights one problem: we can only detect usage of deprecated hooks if the code which invokes the hook is analyzed.